### PR TITLE
Convert all dataclasses to pydantic basemodels

### PR DESCRIPTION
**Pull request by jmorel**
Partial fix of #89 

Replaces all uses of dataclasses in our factory with pydantic BaseModels.

This highlighted several issues in the type hints of Spec objects that I fixed as well.

### CODE REVIEW

File: substratest/factory.py [hunk 1]
```diff
@@ -196,126 +191,113 @@ def read_description(self):
...
+    out_trunk_model_permissions: Permissions
 
     @property
     def id(self):
         return self.composite_traintuple_id
 
 
-@dataclasses.dataclass
-class ComputePlanTesttupleSpec:
+class ComputePlanTesttupleSpec(pydantic.BaseModel):
```
Comment (samlesu) [inline substratest/factory.py:300]:
```suggestion
class ComputePlanTesttupleSpec(_Spec):
```

To be consistent with other classes.

File: substratest/factory.py [hunk 2]
```diff
@@ -196,126 +191,113 @@ def read_description(self):
...
 class TesttupleSpec(_Spec):
     objective_key: str
     traintuple_key: str
-    tag: str
+    tag: str = None
 
 
-@dataclasses.dataclass
-class ComputePlanTraintupleSpec:
+class ComputePlanTraintupleSpec(pydantic.BaseModel):
```
Comment (samlesu) [inline substratest/factory.py:260]:
```suggestion
class ComputePlanTraintupleSpec(_Spec):
```

**Comment by jmorel** [comment]
Thanks for the suggestion, I included it in my commit!

**Comment by jmorel** [comment]
I had to rebase the PR, could you re-approve it?

**Pull request merged**